### PR TITLE
Store IP address in TcpClient

### PR DESCRIPTION
**Pull request by wolfreak99**
This helps address the need for logging ip addresses that join onto servers by logging it into the external console via the server side. This will not show ip addresses for non-host users. In the event that #4273 has confusion with retrieving ip addresses, this will save him the stress and struggles i did back when i needed to figure out how to retrieve ips (as in, googling for about 3 or 4 weeks).

### CODE REVIEW

**Comment by Margen67** [comment]
Would it be possible to also log chat in the console?

File: src/network/TcpSocket.cpp [hunk 1]
```diff
@@ -403,6 +411,16 @@ class TcpSocket : public ITcpSocket
         SDL_UnlockMutex(_connectMutex);
     }
 
+    const char * GetHostStr() override
```
Comment (janisozaur) [inline src/network/TcpSocket.cpp]: this should be const method

Comment (wolfreak99) [inline src/network/TcpSocket.cpp]: huh?

**Comment by wolfreak99** [comment]
@Margen67 
> Would it be possible to also log chat in the console?

It would, but that will be for another pull request feature. I could see that being useful for headless mode or something, But this was meant to remain as minimal as possible that way it wouldn't cause much problems and would help the dude creating the logs.

File: src/network/TcpSocket.cpp (hunk 1 above)
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: https://isocpp.org/wiki/faq/const-correctness

File: src/network/TcpSocket.cpp [hunk 2]
```diff
@@ -201,8 +205,12 @@ class TcpSocket : public ITcpSocket
             }
             else
             {
+                char hoststr[NI_MAXHOST];
+                int rc = getnameinfo((struct sockaddr *)&client_addr, client_len, hoststr, sizeof(hoststr), nullptr, 0, NI_NUMERICHOST | NI_NUMERICSERV);
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Wrap this line like this:
```
int rc = getnameinfo(
    (struct sockaddr *)&client_addr,
    client_len,
    hoststr,
    sizeof(hoststr),
    nullptr,
    0,
    NI_NUMERICHOST | NI_NUMERICSERV);
```

File: src/network/TcpSocket.cpp [hunk 3]
```diff
@@ -201,8 +205,12 @@ class TcpSocket : public ITcpSocket
...
                 SetTCPNoDelay(socket, true);
                 tcpSocket = new TcpSocket(socket);
+                if (rc == 0)
+                    tcpSocket->SetHostStr(hoststr);
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Use braces

File: src/network/TcpSocket.cpp (hunk 1 above)
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Rename this to `const char * GetHostName() const override`

File: src/network/TcpSocket.cpp [hunk 4]
```diff
@@ -403,6 +411,16 @@ class TcpSocket : public ITcpSocket
...
 
+    const char * GetHostStr() override 
+    {
+        return _hostStr.empty() ? nullptr : _hostStr.c_str();
+    }
+
+    void SetHostStr(const char * hostStr) override
+    {
+        _hostStr = std::string(hostStr);
+    }
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Remove this method, its not necessary and certainly not part of the interface.

File: src/network/network.cpp [hunk 5]
```diff
@@ -1408,6 +1408,10 @@ void Network::Server_Client_Joined(const char* name, const std::string &keyhash,
 	if (player) {
 		char text[256];
 		const char * player_name = (const char *) player->name.c_str();
+		//Relay connection information to console
+		auto hostStr = connection.Socket->GetHostStr();
+		log_info("%s has joined from host: %s", player_name, hostStr);
+		//Show join message in chat
```
Comment (IntelOrca) [inline src/network/network.cpp]: Remove this, we will do that actual logging in a separate PR.

File: src/network/TcpSocket.cpp [hunk 6]
```diff
@@ -87,6 +87,8 @@ class TcpSocket : public ITcpSocket
     uint16          _listeningPort  = 0;
     SOCKET          _socket         = INVALID_SOCKET;
 
+    std::string     _hostStr;
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Rename to `_hostName`

File: src/network/TcpSocket.cpp [hunk 7]
```diff
@@ -201,8 +205,12 @@ class TcpSocket : public ITcpSocket
             }
             else
             {
+                char hoststr[NI_MAXHOST];
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Rename to `hostName`

File: src/network/TcpSocket.cpp [hunk 8]
```diff
@@ -201,8 +205,20 @@ class TcpSocket : public ITcpSocket
...
+                    (struct sockaddr *)&client_addr,
+                    client_len,
+                    hostName,
+                    sizeof(hostName),
+                    nullptr,
+                    0,
+                    NI_NUMERICHOST | NI_NUMERICSERV); 
                 SetTCPNoDelay(socket, true);
                 tcpSocket = new TcpSocket(socket);
+                if (rc == 0) {
```
Comment (IntelOrca) [inline src/network/TcpSocket.cpp]: Braces should be on their own lines in C++

File: src/network/TcpSocket.h [hunk 9]
```diff
@@ -58,6 +58,8 @@ interface ITcpSocket
 
     virtual void Disconnect() abstract;
     virtual void Close() abstract;
+
+    virtual const char * GetHostName() const abstract;
```
Comment (IntelOrca) [inline src/network/TcpSocket.h]: Move this to below GetError() for consistency
